pkg/core/dao/cacheddao.go

```diff
@@ -200,5 +211,111 @@ func (cd *Cached) GetWrapped() DAO {
 		cd.unspents,
 		cd.balances,
 		cd.transfers,
+		cd.storage,
 	}
 }
+
+// FlushStorage flushes storage changes to the underlying DAO.
+func (cd *Cached) FlushStorage() error {
+	for h, items := range cd.storage.st {
+		for k, ti := range items {
+			switch ti.State {
+			case putOp:
+				err := cd.DAO.PutStorageItem(h, []byte(k), &ti.StorageItem)
+				if err != nil {
+					return err
+				}
+			case delOp:
+				err := cd.DAO.DeleteStorageItem(h, []byte(k))
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// GetStorageItem returns StorageItem if it exists in the given store.
+func (cd *Cached) GetStorageItem(scripthash util.Uint160, key []byte) *state.StorageItem {
+	ti := cd.storage.getItem(scripthash, key)
+	if ti != nil {
+		if ti.State == delOp {
+			return nil
+		}
+		return &ti.StorageItem
+	}
+
+	si := cd.DAO.GetStorageItem(scripthash, key)
+	if si != nil {
+		cd.storage.put(scripthash, key, &trackedItem{
+			StorageItem: *si,
+			State:       getOp,
+		})
+	}
+	return si
+}
+
+// PutStorageItem puts given StorageItem for given script with given
+// key into the given store.
+func (cd *Cached) PutStorageItem(scripthash util.Uint160, key []byte, si *state.StorageItem) error {
+	val := make([]byte, len(si.Value))
+	copy(val, si.Value)
+	cd.storage.put(scripthash, key, &trackedItem{
+		StorageItem: state.StorageItem{
+			Value:   val,
+			IsConst: si.IsConst,
+		},
+		State: putOp,
+	})
+	return nil
+}
+
+// DeleteStorageItem drops storage item for the given script with the
+// given key from the store.
+func (cd *Cached) DeleteStorageItem(scripthash util.Uint160, key []byte) error {
+	cd.storage.put(scripthash, key, &trackedItem{
+		State: delOp,
+	})
+	return nil
+}
+
+// GetStorageItems returns all storage items for a given scripthash.
+func (cd *Cached) GetStorageItems(hash util.Uint160) ([]StorageItem, error) {
+	items, err := cd.DAO.GetStorageItems(hash)
+	if err != nil {
+		return nil, err
+	}
+
+	cache := cd.storage.getItems(hash)
+	if len(cache) == 0 {
+		return items, nil
+	}
+	result := make([]StorageItem, 0, len(items))
+	for i := range items {
+		_, ok := cache[string(items[i].Key)]
+		if !ok {
+			result = append(result, items[i])
+		}
+	}
+
+	l := len(result)
+	for k, v := range cache {
+		if v.State != delOp {
+			val := make([]byte, len(v.StorageItem.Value))
+			copy(val, v.StorageItem.Value)
+			result = append(result, StorageItem{
+				StorageItem: state.StorageItem{
+					Value:   val,
+					IsConst: v.StorageItem.IsConst,
+				},
+				Key: []byte(k),
+			})
+		}
+	}
+
+	s := result[l:]
+	sort.Slice(s, func(i, j int) bool { return bytes.Compare(s[i].Key, s[j].Key) == 1 })
```

Comment:
IIUC we should have the following order:
* cached items, ordered by the time of their addition to cache (like maps)
* items from storage naturally ordered by keys, except ones that were cached

Reply:
Fixed.